5 Object Oriented Programming/OOP_readme.md

```diff
@@ -0,0 +1,141 @@
+
+
+### **Inheritance**
+
+JavaScript's prototype system makes it possible to create a new class, much like the old class, but with new definitions for some of its properties. The prototype for the new class derives from the old prototype but adds a new definition for, say, the set method. In object-oriented programming terms, this is called inheritance. The new class inherits properties and behaviour from the old class.
+```javascript
+class SymmetricMatrix extends Matrix {
+	        super(size, size, (x, y) => {
+            if (x < y) return element(y, x);
+            else return element(x, y);
+        });
+    }
+    set(x, y, value) {
+        super.set(x, y, value);
+        if (x!= y) {
+            super.set(y, x, value);
+});
+}
+	set(x, y, value) {
+	super.set(x, y, value);
+	if (x!= y) {
+	super.set(y, x, value);
+	}
+}}
+
+let matrix = new SymmetricMatrix(5, (x, y) => `${x},${y}`);
+
+console.log(matrix.get(2, 3)); //3,2
+```
+
+The use of the word extends indicates that this class shouldn't be directly based on the default object prototype but on some other class. This class is called the superclass. The derived class is called subclass. To initialize a SymmetricMatrix instance, the constructor calls its superclass's constructor through the super keyword. This is necessary because if this new object is to behave like a Matrix, it is going to need the instance properties that matrices have. To ensure the matrix is symmetrical, the constructor wraps the element function to swap the coordinates for values below the diagonal. The set method again uses super but this time not to call the constructor but to call a specific method from the superclass's set of methods. Inside class methods, super provides a way to call methods as they were defined in the superclass. Inheritance allows us to build slightly different data types from existing data types with relatively little work. It is a fundamental part of the object-oriented tradition, alongside encapsulation and polymorphism.
+
+#### **The instanceof operator**
+
+It is occasionally useful to know whether an object was derived from a specific class. For this, JavaScript provides a binary operator called instanceof.
+```javascript
+console.log( new SymmetricMatrix(2) instanceof SymmetricMatrix); // true
+console.log(new SymmetricMatrix(2) instanceof Matrix); // true
+console.log(new Matrix(2, 2) instanceof SymmetricMatrix); // false
+console.log([1] instanceof Array); // true
+```
+The operator will see through inherited types, so a SymmetricMatrix is an instance of Matrix. The operator can also be applied to standard constructors like Array. Almost every object is an instance of Object.

```

Comment:
The format and explanation look good! 👍🏼

Reply:
Thanks a lot!